jib-core/src/test/java/com/google/cloud/tools/jib/event/progress/AllocationTest.java

```diff
@@ -0,0 +1,93 @@
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.event.progress;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+/** Tests for {@link Allocation}. */
+public class AllocationTest {
+
+  /** Error margin for checking equality of two doubles. */
+  private static final double DOUBLE_ERROR_MARGIN = 1e-10;
+
+  @Test
+  public void testSmoke_linear() {
+    Allocation root = Allocation.newRoot("root", 1);
+    Allocation node1 = root.newChild("node1", 2);
+    Allocation node2 = node1.newChild("node2", 3);
+
+    Assert.assertEquals("node2", node2.getDescription());
+    Assert.assertEquals(3, node2.getAllocationUnits());
+    Assert.assertEquals(1.0 / 2, node2.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertTrue(node2.getParent().isPresent());
+    Assert.assertEquals(node1, node2.getParent().get());
+
+    Assert.assertEquals("node1", node1.getDescription());
+    Assert.assertEquals(2, node1.getAllocationUnits());
+    Assert.assertTrue(node1.getParent().isPresent());
+    Assert.assertEquals(root, node1.getParent().get());
+    Assert.assertEquals(1.0, node1.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+
+    Assert.assertEquals("root", root.getDescription());
+    Assert.assertEquals(1, root.getAllocationUnits());
+    Assert.assertFalse(root.getParent().isPresent());
+    Assert.assertEquals(1.0, root.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+  }
+
+  @Test
+  public void testFractionOfRoot_tree_partial() {
+    Allocation root = Allocation.newRoot("ignored", 10);
+    Allocation left = root.newChild("ignored", 2);
+    Allocation right = root.newChild("ignored", 4);
+    Allocation leftDown = left.newChild("ignored", 20);
+    Allocation rightLeft = right.newChild("ignored", 20);
+    Allocation rightRight = right.newChild("ignored", 100);
+    Allocation rightRightDown = rightRight.newChild("ignored", 200);
+
+    Assert.assertEquals(1.0, root.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(1.0 / 10, left.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(1.0 / 10, right.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(1.0 / 10 / 2, leftDown.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(1.0 / 10 / 4, rightLeft.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(1.0 / 10 / 4, rightRight.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+    Assert.assertEquals(
+        1.0 / 10 / 4 / 100, rightRightDown.getFractionOfRoot(), DOUBLE_ERROR_MARGIN);
+  }
+
+  @Test
+  public void testFractionOfRoot_tree_complete() {
```

Comment:
So I see this example shows that the expect normal usage is to match the number of children and the parent allocation units. As I said in my [previous comment](https://github.com/GoogleContainerTools/jib/pull/1298#issuecomment-444320149), I am really curious how this can pan out or be enforced efficiently in practice. But I think we can move forward with this now, and I'd be able to see if the API will work out to minimize errors.

Reply:
Yea, the user of the `Allocation` class will be responsible for sub-allocating the correct number of allocations. For example, for a task that knows it will launch 3 subtasks, it will first create its own allocation as 3 allocation units, and then launch the 3 subtasks, passing down its own allocation as the "parent" allocation, knowing that each of the subtasks will thus make a suballocation, resulting in 3 total suballocations.

Reply:
Yeah, in that example, what if the writer of one of the subtasks (think of as separate methods) that get the "parent" allocation doesn't have anything to report progress or if the writer simply wants to mark completion before returning using the given parent allocation node? Should it emit an event with the parent node and 1 allocation done? But, in turn, what if the subtask called other methods and propagated down the parent node as-is? In that case, if the sub-subtasks did the same thing as what the subtask did (emitting the event with the parent node and 1 allocation done), the progress will go over the allocated amount.

Reply:
Yes, emitted an event with the parent node and 1 allocation done can work, or the subtask can create a suballocation and emit an event with all progress units in the suballocation. The allocation(s) are only to represent the subdivision of the overall progress and it will be up to the tasks to coordinate how to emit progress on those allocations (with the premise that parent tasks have power over child tasks) and up to the progress event listener to process the emitted progress in whichever way it chooses to do so (allow or disallow going over allocated amount, for eg.).

Reply:
> with the premise that parent tasks have power over child tasks)

Yeah, I think this premise is required for this work well. When the premise is applied recursively, I think it implies having power over all descendants. For example, if you are some subtask at some point, and you call several long-running external methods in sequence, you don't really know about whose internals. So, you just pass the allocation node you have, but you basically have no idea how those will make use up the node. They may or may not consume 1 allocation, or more. And those external methods may in the same situation recursively. So, basically, it looks to me that we should have the premise that parent tasks should have an idea about all descendants.

Reply:
Yeah, in our use case in the builder steps, this premise will be met by a constraint that 

1. A task may only create at most one suballocation of the parent allocation passed to it, and
1. A task cannot pass the parent allocation on to subtasks